extensions/amp-ad/0.1/amp-ad-batch-manager.js

```diff
@@ -0,0 +1,165 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {user} from '../../../src/log';
+import {xhrFor} from '../../../src/xhr';
+
+/** @const {number} The number of milliseconds between checks to see if the response has arrived */
+const AD_BATCH_TICK_LENGTH = 50;
+
+/** @const {number} The number of ticks to wait for a response from the ad server before timing out */
+const AD_BATCH_TIMEOUT = 100;
+
+/** @const {!string} Tag name for ad batch manager implementation. */
+export const TAG_AD_BATCH_MANAGER = 'amp-ad-batch-manager';
+
+/**
+  * Get a batch manager that will be used to batch together elements with the same data-url
+  * Note a side effect: the first ad with a given data-url to be processed will have its isBatchMaster_ variable set to true
+  * @param {Element} e The element which is to be batched
+  * @param {string} url The URL identifying which batch
+  * @param {string} selector The selector to choose the things to be batched (e.g. 'amp-ad[type=imagead]')
+  * @returns {AmpAdBatchManager}
+  */
+export function getBatchManager(e, url, selector) {
+  // If this is our first imagead, create a map of responses for each value of 'data-url'
+  // This allows ads from multiple ad servers on the same page
+  if (!(e.win.hasOwnProperty('imageadBatchManagers'))) {
+    // Create an array of batch mangers. There will be one for each URL.
+    e.win.imageadBatchManagers = {};
+    // Scan the page to get a list of all image ads on this page, indexed by url.
+    // For performance reasons, we only do the scan once, the first time that a batch manager is constructed.
+    e.win.imageadElements = {};
+    const elements = document.querySelectorAll(selector);
+    for (let i = 0; i < elements.length; i++) {
+      const el = elements[i];
+      const url = el.getAttribute('data-url');
+      if (!(url in e.win.imageadElements)) {
+        e.win.imageadElements[url] = [];
+      }
+      e.win.imageadElements[url].push(el);
+    }
+  }
+  if (!(url in e.win.imageadBatchManagers)) {
+    // This is the first time we've seen this URL, so this will be the master for this batch
+    e.isBatchMaster_ = true;
```

Comment:
Possibly, but the calling code then gets very muddy, since you only have a promise of a batch manager, not an actual batch manager. I guess I could investigate, but it feels like a lot of work.